tests/src/Interop/ArrayMarshalling/SafeArray/SafeArrayTest.cs

```diff
@@ -0,0 +1,219 @@
+
+            var objects = new object[] { new object(), new object(), new object() };
+            Assert.IsTrue(SafeArrayNative.VerifyIUnknownArray(objects));
+            Assert.IsTrue(SafeArrayNative.VerifyIDispatchArray(objects));
+
+            var variantInts = new object[] {1, 2, 3, 4, 5, 6, 7, 8, 9};
+            
+            Assert.IsTrue(SafeArrayNative.MeanVariantIntArray(variantInts, out var variantMean));
+            Assert.AreEqual(variantInts.OfType<int>().Average(), variantMean);
+
+            var dates = new DateTime[] { new DateTime(2008, 5, 1), new DateTime(2010, 1, 1) };
+            Assert.IsTrue(SafeArrayNative.DistanceBetweenDates(dates, out var numDays));
+            Assert.AreEqual((dates[1] - dates[0]).TotalDays, numDays);
+
+            Assert.IsTrue(SafeArrayNative.XorBoolArrayInStruct(
+                new SafeArrayNative.StructWithSafeArray
+                {
+                    values = boolArray
+                },
+                out var structXor));
+
+            Assert.AreEqual(XorArray(boolArray), structXor);
+        }
+        catch (Exception e)
+        {
+            Console.WriteLine(e);
+            return 101;
+        }
+        return 100;
+    }
+
+    private static bool XorArray(bool[] values)
+    {
+        bool retVal = false;
+        foreach (var item in values)
+        {
+            retVal ^= item;
+        }
+        return retVal;
+    }
+
+    private static string Reverse(string s)
+    {
+        var chars = s.ToCharArray();
+        Array.Reverse(chars);
+        return new string(chars);
+    }
+}
+
+class SafeArrayNative
+{
+    public struct StructWithSafeArray
+    {
+        [MarshalAs(UnmanagedType.SafeArray, SafeArraySubType = VarEnum.VT_BOOL)]
+        public bool[] values;
+    }
+
+    public struct BlittableRecord
+    {
+        public int a;
+    }
+
+    public struct NonBlittableRecord
+    {
+        public bool b;
+    }
+
+    [DllImport(nameof(SafeArrayNative))]
+    [return: MarshalAs(UnmanagedType.SafeArray, SafeArraySubType = VarEnum.VT_RECORD)]
+    private static extern  BlittableRecord[] CreateSafeArrayOfRecords(
+        BlittableRecord[] records,
+        int numElements
+    );
+
+    public static BlittableRecord[] CreateSafeArrayOfRecords(BlittableRecord[] records)
+    {
+        return CreateSafeArrayOfRecords(records, records.Length);
+    }
+
+    [DllImport(nameof(SafeArrayNative), EntryPoint = "CreateSafeArrayOfNonBlittableRecords")]
+    [return: MarshalAs(UnmanagedType.SafeArray, SafeArraySubType = VarEnum.VT_RECORD)]
+    private static extern NonBlittableRecord[] CreateSafeArrayOfRecords(
+        NonBlittableRecord[] records,
+        int numElements
+    );
+
+    public static NonBlittableRecord[] CreateSafeArrayOfRecords(NonBlittableRecord[] records)
+    {
+        return CreateSafeArrayOfRecords(records, records.Length);
+    }
+
+    [DllImport(nameof(SafeArrayNative))]
+    public static extern bool XorBoolArray(
```

Comment:
All of the associated `out` params can now just be natural return values.

Reply:
For an example that fails:
```
    [DllImport(nameof(SafeArrayNative), PreserveSig = false)]
    public static extern decimal MeanDecimalArray(
        [MarshalAs(UnmanagedType.SafeArray)] decimal[] values);
```

Reply:
Using `[return:MarshalAs(…)]` syntax doesn't do the HResult swapping that would allow us to use an out parameter as a return value.